From 4fb186ca9e0af4bc7e10af2e915485caa0d6b1c0 Mon Sep 17 00:00:00 2001
From: Ivan Hu <ivan.hu@canonical.com>
Date: Thu, 14 Nov 2019 15:06:30 +0800
Subject: [PATCH 2/2] device-drivers/acpi/ltp_acpi_cmd: skip pm2 register test
 if not supported

Some platforms got test fail on ltp_acpi test-case 8,
ltp_acpi 8 TPASS : Test-case '7'
ltp_acpi 9 TFAIL : ltp_acpi.c:139: Test-case '8'
ltp_acpi 10 TPASS : Test-case '9'

[ 504.084067] ltp_acpi_test: get register: 10 val: 0000
[ 504.084541] ltp_acpi_test: get register: 11 val: 0000
[ 504.084944] ltp_acpi_test: get register: 12 val: 0000
[ 504.085318] ACPI Exception: AE_BAD_ADDRESS, acpi_read_bit_register (20170831/LTP_ACPI-61)

The failure is on reading ID 0x13, PM2 control bit 0(ARB_DIS), and getting bad
address return, caused by PM2 isn't supported by firmware.

This patch checks the PM2 block on FADT table, if not exist, skip the testing.

Signed-off-by: Ivan Hu <ivan.hu@canonical.com>
---
 testcases/kernel/device-drivers/acpi/ltp_acpi_cmds.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/testcases/kernel/device-drivers/acpi/ltp_acpi_cmds.c b/testcases/kernel/device-drivers/acpi/ltp_acpi_cmds.c
index 2733753cc..0ce70c6c1 100644
--- a/testcases/kernel/device-drivers/acpi/ltp_acpi_cmds.c
+++ b/testcases/kernel/device-drivers/acpi/ltp_acpi_cmds.c
@@ -233,6 +233,9 @@ static int acpi_traverse_from_root(void)
 static acpi_handle dev_handle;
 static int acpi_hw_reduced;
 
+/* check if PM2 control register supported */
+static bool pm2_supported;
+
 static int acpi_init(void)
 {
 	acpi_status status;
@@ -240,12 +243,15 @@ static int acpi_init(void)
 	struct acpi_table_fadt *fadt;
 	struct acpi_table_header *table;
 	struct acpi_device_info *dev_info;
+	pm2_supported = true;
 
 	status = acpi_get_table(ACPI_SIG_FADT, 0, &table);
 	if (ACPI_SUCCESS(status)) {
 		fadt = (struct acpi_table_fadt *)table;
 		if (fadt->flags & ACPI_FADT_HW_REDUCED)
 			acpi_hw_reduced = 1;
+		if (fadt->pm2_control_block == 0 || fadt->pm2_control_length == 0)
+			pm2_supported = false;
 	}
 	if (acpi_hw_reduced)
 		prk_alert("Detected the Hardware-reduced ACPI mode");
@@ -481,6 +487,8 @@ static int acpi_test_register(void)
 	 * commit/?id=50ffba1bd3120b069617455545bc27bcf3cf7579
 	 */
 	for (i = 0; i < ACPI_NUM_BITREG; ++i) {
+		if (i == ACPI_BITREG_ARB_DISABLE && !pm2_supported)
+			continue;
 		status = acpi_read_bit_register(i, &val);
 		err |= acpi_failure(status, "acpi_read_bit_register");
 		if (ACPI_SUCCESS(status))
-- 
2.17.1

